webapp/src/ts/services/transitions/create-user-for-contacts.transition.ts

```diff
@@ -0,0 +1,173 @@
+    }
+
+    return docs;
+  }
+
+  private getTransitionConfig(settings) {
+    return settings[this.name];
+  }
+
+  private getReplaceForms(settings) {
+    return this.getTransitionConfig(settings)?.[this.REPLACE_PROPERTY] || [];
+  }
+
+  private getUserReplaceDoc(docs: Doc[]) {
+    return docs.find(doc => this.replaceForms.includes(doc.form)) as UserReplaceDoc;
+  }
+
+  private async replaceUser(docs: Doc[], userReplaceDoc: UserReplaceDoc, originalContact: Doc) {
+    const { replacement_contact_id } = userReplaceDoc.fields;
+    if (!replacement_contact_id) {
+      throw new Error('The form for replacing a user must include a replacement_contact_id field ' +
+        'containing the id of the new contact.');
+    }
+
+    const originalContactId = userReplaceDoc.contact._id;
+
+    const isOriginalContact = () => originalContact._id === originalContactId;
+    const isReplacedContact = () => this.createUserForContactsService.isReplaced(originalContact) &&
+      this.createUserForContactsService.getReplacedBy(originalContact) === originalContactId;
+    if (!isOriginalContact() && !isReplacedContact()) {
+      throw new Error('Only the contact associated with the currently logged in user can be replaced.');
+    }
+    const newContact = await this.getNewContact(docs, replacement_contact_id);
+    if (!newContact) {
+      throw new Error(`The new contact could not be found [${replacement_contact_id}].`);
+    }
+    this.createUserForContactsService.setReplaced(originalContact, newContact);
+    const updatedDocs = [...docs, originalContact];
+
+    const parentPlace = await this.getParentDoc(newContact);
+    if (parentPlace?.contact?._id === originalContactId) {
+      parentPlace.contact = this.extractLineageService.extract(newContact);
+      updatedDocs.push(parentPlace);
+    }
+
+    return updatedDocs;
+  }
+
+  private async getParentDoc(doc: ContactDoc) {
+    return this.dbService
+      .get()
+      .get(doc.parent._id)
+      .catch(err => {
+        if (err.status === 404) {
+          return;
+        }
+        throw err;
+      }) as ContactDoc;
+  }
+
+  private async getNewContact(docs: Doc[], newContactId: string): Promise<ContactDoc> {
+    const newContact = docs.find(doc => doc._id === newContactId);
+    if (newContact) {
+      return newContact as ContactDoc;
+    }
+    return this.dbService
+      .get()
+      .get(newContactId)
+      .catch(err => {
+        if (err.status === 404) {
+          return;
+        }
+        throw err;
+      });
+  }
+
+  private getReportDocs(docs: Doc[]) {
+    return docs
+      .filter(doc => doc.type === 'data_record')
+      .filter(doc => doc.contact) as ReportDoc[];
+  }
+
+  private reparentReports(docs: Doc[], originalContact: Doc) {
+    const replacedById = this.createUserForContactsService.getReplacedBy(originalContact);
+    if (!replacedById) {
+      return;
+    }
+    this.getReportDocs(docs)
```

Comment:
I have refactored this so that we iterate twice: once to filter and once to mutate.  IMHO the readability benefits of separating the filtering from the mutating outweigh the slight performance gains from combining them.

Reply:
Yes, definitely, since we know this is going to be a small-ish array. If we'd have a 1kk length array, I would definitely go for the combined approach.